# Import library and read dummy dataset

In [1]:
import pandas as pd
import numpy as np
dataframe = pd.read_csv('data.csv') # Dummy dataset containing 50 rows.

In [2]:
dataframe

R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80        471784.10    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6   134615.46       147198.87              NaN  California  156122.51
7   130298.13       145530.06        323876.68         NaN  155752.60
8   120542.52       148718.95        311613.29    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       135495.07        252664.93  California  134307.35
14  119943.24       156547.42        256512.92     Florida  132602.65
15  114523.61       122616.84        261776.23         NaN  129917.04
16   78013.11       121597.55        264346.06  California  126992.93
17   94657.16       145077.58        282574.31    New York  125370.37
18   91749.16       114175.79        294919.57     Florida  124266.90
19   86419.70       153514.11             0.00         NaN  122776.86
20   76253.86       113867.30        298664.47  California  118474.03
21   78389.47       153773.43        299737.29    New York  111313.02
22   73994.56       122782.75        303319.26     Florida  110352.25
23   67532.53       105751.03        304768.73         NaN  108733.99
24   77044.01        99281.34        140574.81    New York  108552.04
25   64664.71       139553.16        137962.62  California  107404.34
26   75328.87       144135.98              NaN     Florida  105733.54
27   72107.60       127864.55        353183.81    New York  105008.31
28   66051.52       182645.56        118148.20     Florida  103282.38
29   65605.48       153032.06        107138.38         NaN  101004.64
30   61994.48       115641.28         91131.24     Florida   99937.59
31   61136.38       152701.92              NaN    New York   97483.56
32   63408.86       129219.61         46085.25  California   97427.84
33   55493.95       103057.49        214634.81     Florida   96778.92
34   46426.07       157693.92        210797.67  California   96712.80
35   46014.02        85047.44        205517.64    New York   96479.51
36   28663.76       127056.21        201126.82     Florida   90708.19
37   44069.95        51283.14        197029.42  California   89949.14
38   20229.59        65947.93        185265.10    New York   81229.06
39   38558.51        82982.09        174999.30  California   81005.76
40   28754.33       118546.05        172795.67  California   78239.91
41   27892.92        84710.77              NaN     Florida   77798.83
42   23640.93        96189.63        148001.11  California   71498.49
43   15505.73       127382.30         35534.17    New York   69758.98
44   22177.74       154806.14         28334.72  California   65200.33
45    1000.23       124153.04          1903.93    New York   64926.08
46    1315.46       115816.21        297114.46     Florida   49490.75
47       0.00       135426.92             0.00  California   42559.73
48     542.05        51743.15             0.00    New York   35673.41
49       0.00       116983.80         45173.06  California   14681.40

# 5

In [3]:
dataframe.isnull().values.any() # To check dataframe for null Value

True

In [4]:
dataframe.isna().any() # Check which column contains NaN value

R&D Spend          False
Administration     False
Marketing Spend     True
State               True
Profit             False
dtype: bool

# 6

In [5]:
dataframe.columns[dataframe.isna().any()].tolist() # column name containing NaN as list

['Marketing Spend', 'State']

In [6]:
# check for categorical or Numerical data
dataframe.head() # display top 5 rows

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [7]:
dataframe.dtypes

R&D Spend          float64
Administration     float64
Marketing Spend    float64
State               object
Profit             float64
dtype: object

In [8]:
# as we can see in the above table that column 'Marketing Spend' contains Numerical data 
# and 'State' contains categorical data.

# 7
- Replace NaN to Most common value in case of categorical data.

In [9]:
# Value count : to know most common value in categorical data
dataframe['State'].value_counts()

California    17
Florida       14
New York      14
Name: State, dtype: int64

In [10]:
most_common = dataframe['State'].value_counts(sort=True).index[0] #get most common value
most_common

'California'

In [11]:
# Replace NaN to Most common value in case of categorical data.
dataframe.fillna({'State':most_common},inplace=True)

- Replace NaN to average of all the other values in case of Numerical data

In [12]:
# For numerical data
avg = np.mean(dataframe['Marketing Spend'])
avg

218191.2839130435

In [13]:
dataframe.fillna({'Marketing Spend':avg},inplace=True) 

# 8

In [14]:
dataframe.isnull().values.any() # no NaN exists in entire dataframe

False

In [15]:
dataframe.isna().any() # check in all columns.

R&D Spend          False
Administration     False
Marketing Spend    False
State              False
Profit             False
dtype: bool

# 9

In [16]:
dataframe.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [17]:
# y : last column (4th column) i.e profit
# dataframe.iloc[rows,columns]
# :   :- all ,eg all rows
# [3] :- 3rd column
# [:3] :-column 0,1 and 2
y = dataframe.iloc[:,[4]] 
x_categorical = dataframe.iloc[:,[3]] # categorical data in 3rd column
x_numerical = dataframe.iloc[:,:3] # Numerical data in column-0 ,column-1,column-2.

In [18]:
y.head() #display first 5 rows

Profit
0  192261.83
1  191792.06
2  191050.39
3  182901.99
4  166187.94

In [19]:
x_categorical.head() #display first 5 rows

State
0    New York
1  California
2     Florida
3    New York
4     Florida

In [20]:
x_numerical.head() #display first 5 rows

R&D Spend  Administration  Marketing Spend
0  165349.20       136897.80        471784.10
1  162597.70       151377.59        443898.53
2  153441.51       101145.55        407934.54
3  144372.41       118671.85        383199.62
4  142107.34        91391.77        366168.42

# 10

In [88]:
dataframe

R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80    471784.100000    New York  192261.83
1   162597.70       151377.59    443898.530000  California  191792.06
2   153441.51       101145.55    407934.540000     Florida  191050.39
3   144372.41       118671.85    383199.620000    New York  182901.99
4   142107.34        91391.77    366168.420000     Florida  166187.94
5   131876.90        99814.71    362861.360000    New York  156991.12
6   134615.46       147198.87    218191.283913  California  156122.51
7   130298.13       145530.06    323876.680000  California  155752.60
8   120542.52       148718.95    311613.290000    New York  152211.77
9   123334.88       108679.17    304981.620000  California  149759.96
10  101913.08       110594.11    229160.950000     Florida  146121.95
11  100671.96        91790.61    249744.550000  California  144259.40
12   93863.75       127320.38    249839.440000     Florida  141585.52
13   91992.39       135495.07    252664.930000  California  134307.35
14  119943.24       156547.42    256512.920000     Florida  132602.65
15  114523.61       122616.84    261776.230000  California  129917.04
16   78013.11       121597.55    264346.060000  California  126992.93
17   94657.16       145077.58    282574.310000    New York  125370.37
18   91749.16       114175.79    294919.570000     Florida  124266.90
19   86419.70       153514.11         0.000000  California  122776.86
20   76253.86       113867.30    298664.470000  California  118474.03
21   78389.47       153773.43    299737.290000    New York  111313.02
22   73994.56       122782.75    303319.260000     Florida  110352.25
23   67532.53       105751.03    304768.730000  California  108733.99
24   77044.01        99281.34    140574.810000    New York  108552.04
25   64664.71       139553.16    137962.620000  California  107404.34
26   75328.87       144135.98    218191.283913     Florida  105733.54
27   72107.60       127864.55    353183.810000    New York  105008.31
28   66051.52       182645.56    118148.200000     Florida  103282.38
29   65605.48       153032.06    107138.380000  California  101004.64
30   61994.48       115641.28     91131.240000     Florida   99937.59
31   61136.38       152701.92    218191.283913    New York   97483.56
32   63408.86       129219.61     46085.250000  California   97427.84
33   55493.95       103057.49    214634.810000     Florida   96778.92
34   46426.07       157693.92    210797.670000  California   96712.80
35   46014.02        85047.44    205517.640000    New York   96479.51
36   28663.76       127056.21    201126.820000     Florida   90708.19
37   44069.95        51283.14    197029.420000  California   89949.14
38   20229.59        65947.93    185265.100000    New York   81229.06
39   38558.51        82982.09    174999.300000  California   81005.76
40   28754.33       118546.05    172795.670000  California   78239.91
41   27892.92        84710.77    218191.283913     Florida   77798.83
42   23640.93        96189.63    148001.110000  California   71498.49
43   15505.73       127382.30     35534.170000    New York   69758.98
44   22177.74       154806.14     28334.720000  California   65200.33
45    1000.23       124153.04      1903.930000    New York   64926.08
46    1315.46       115816.21    297114.460000     Florida   49490.75
47       0.00       135426.92         0.000000  California   42559.73
48     542.05        51743.15         0.000000    New York   35673.41
49       0.00       116983.80     45173.060000  California   14681.40

In [89]:
X = dataframe.iloc[:,:4] # splitting dataset for X value

In [90]:
X.head()

R&D Spend  Administration  Marketing Spend       State
0  165349.20       136897.80        471784.10    New York
1  162597.70       151377.59        443898.53  California
2  153441.51       101145.55        407934.54     Florida
3  144372.41       118671.85        383199.62    New York
4  142107.34        91391.77        366168.42     Florida

In [91]:
X_num = X.iloc[:,:3] # splitting to numerical value
X_cat = X.iloc[:,3] # splitting to categorical value

In [92]:
X_cat.head()

0      New York
1    California
2       Florida
3      New York
4       Florida
Name: State, dtype: object

In [93]:
X_num.head()

R&D Spend  Administration  Marketing Spend
0  165349.20       136897.80        471784.10
1  162597.70       151377.59        443898.53
2  153441.51       101145.55        407934.54
3  144372.41       118671.85        383199.62
4  142107.34        91391.77        366168.42

# 11

In [94]:
X_cat.value_counts()

California    22
Florida       14
New York      14
Name: State, dtype: int64

In [95]:
dataframe_cat = pd.DataFrame()

In [96]:
# converting categorical value to 1/0 encoding
dataframe_cat['California'] = np.where(X_cat=='California' ,1,0) 
dataframe_cat['New York'] = np.where(X_cat=='New York' ,1,0) 
dataframe_cat['Florida'] = np.where(X_cat=='Florida' ,1,0) 

In [97]:
dataframe_cat.head()

California  New York  Florida
0           0         1        0
1           1         0        0
2           0         0        1
3           0         1        0
4           0         0        1

# 12

In [98]:
# merge X_num and dataframe_cat side by side
glued_dataframe = pd.concat([X_num, dataframe_cat], axis=1)
glued_dataframe.head()

R&D Spend  Administration  Marketing Spend  California  New York  Florida
0  165349.20       136897.80        471784.10           0         1        0
1  162597.70       151377.59        443898.53           1         0        0
2  153441.51       101145.55        407934.54           0         0        1
3  144372.41       118671.85        383199.62           0         1        0
4  142107.34        91391.77        366168.42           0         0        1

In [99]:
X = glued_dataframe

# 13

In [100]:
# training and test set split and model creation
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

# 14

In [101]:
model.coef_ # coefficients

array([[ 7.90910038e-01,  2.96021037e-02,  2.98037742e-02,
        -1.43873763e+03,  6.45911591e+02,  7.92826035e+02]])

In [102]:
# below code was given is question
print('Model Intercept: ',model.intercept_)
coefResults = list(zip(X.columns,model.coef_[0]))
for coefResult in coefResults:
    print(str(coefResult[0]).ljust(30)," ",str(coefResult[1]).rjust(25))

Model Intercept:  [43368.59437609]
R&D Spend                               0.7909100384023781
Administration                        0.029602103683761132
Marketing Spend                        0.02980377419672228
California                              -1438.737626104836
New York                                 645.9115908083481
Florida                                  792.8260352964897


# 15

- here i am interpreting 'R&D Spend'
- 1 unit change in 'R&D Spend' will result in .7909 unit change in independent variable (y)
- i.e overall 'profit' in our dataset

# 16

In [103]:
X.corr() # finding correlation between columns

R&D Spend  Administration  Marketing Spend  California  \
R&D Spend         1.000000        0.241955         0.746604   -0.051707   
Administration    0.241955        1.000000         0.005887    0.092320   
Marketing Spend   0.746604        0.005887         1.000000   -0.207324   
California       -0.051707        0.092320        -0.207324    1.000000   
New York         -0.003269       -0.099167         0.074245   -0.552771   
Florida           0.060433       -0.002897         0.154960   -0.552771   

                 New York   Florida  
R&D Spend       -0.003269  0.060433  
Administration  -0.099167 -0.002897  
Marketing Spend  0.074245  0.154960  
California      -0.552771 -0.552771  
New York         1.000000 -0.388889  
Florida         -0.388889  1.000000

In [104]:
# example. 'California' and 'R&D Spend' are negatively related 
# 'Marketing Spend' and 'R&D Spend' are highly correlated
# similarly you can find the relation in your dataset.